x-pack/docs/en/security/configuring-stack-security.asciidoc

```diff
@@ -112,3 +125,78 @@ include::{es-ref-dir}/setup/install/security-files-reference.asciidoc[leveloffse
+that the node is starting. If any of those checks fail, there's a good
+indication that you <<<<manually-configure-security,manually configured security>>,
+or don't want security to be configured automatically. In these cases, we'll
+respect your existing configuration and skip the auto-configuration process. 
+
+[discrete]
+[[stack-existing-environment-detected]]
+==== Existing environment detected
+If certain directories already exist, there's a strong indication that the node
+was started previously. Similarly, if certain files _don't_ exist, or we can't 
+read or write to specific files or directories, then we're likely not running as 
+the user who installed {es} or an administrator imposed restrictions. If any of 
+the following environment checks are true, security isn't configured 
+automatically.
+
+The {es} `/data` directory exists and isn't empty::
+The existence of this directory is a strong indicator that the node was started 
+previously, and we don't want to overwrite your existing configuration.
+
+The `elasticsearch.yml` file doesn't exist (or isn't readable), or the `elasticsearch.keystore` isn't readable::
+If either of these files aren't readable, we can't determine whether {es} security 
+features are already enabled. This state also indicates that we're probably not 
+running as a user who has sufficient privileges.
+
+The {es} configuration directory isn't writable::
+This state likely indicates that an administrator made this directory read-only,
+or that the user who is starting {es} is not the user that installed {es}.
+
+Indices are read-only::
+It's likely that there's not enough disk space, so we can't write the `elastic`
+user password and API key to the `.security` index.
+
+[discrete]
+[[stack-existing-settings-detected]]
+==== Existing settings detected
+If {es} detects any of the following settings, then we'll assume that you want
+to configure security on your own and we won't configure security automatically.
+
+* {ref}/modules-node.html#node-roles[`node.roles`] is set to a value other than 
+`master`
+* {ref}/security-settings.html#general-security-settings[`xpack.security.autoconfiguration.enabled`] is set to `false`
+* {ref}/security-settings.html#general-security-settings[`xpack.security.enabled`] has a value set
+* Any of the
+{ref}/security-settings.html#transport-tls-ssl-settings[`xpack.security.transport.ssl.*`] or 
+{ref}/security-settings.html#http-tls-ssl-settings[`xpack.security.http.ssl.*`]
+settings have a value set
+* Any of the `discovery.type`, `discovery.seed_hosts`, or
+`cluster.initial_master_nodes`
+{ref}/modules-discovery-settings.html[discovery and cluster formation settings] 
+have a value set
++
+--
+[NOTE]
+====
+Exceptions are when `discovery.type` is set to `single-node`, or when
+`cluster.initial_master_nodes` exists but contains only the name of the current
+node.
+====
+--
+
+[discrete]
+[[stack-existing-keystore-detected]]
+==== Keystore settings detected
```

Comment:
TLDR I would drop this entire section. I think it is too much of an implementation detail.

> we can't generate a password for the `elastic` user

That's not quire true, that is not the reason.
If there are TLS settings in the keystore we don't overwrite them and the node will fail to start.

Reply:
Is it worth changing this section to say:

>If the node startup process detects existing TLS settings in the Elasticsearch keystore, the process halts and the node fails to start.

Do users need to know even that? Or is it still too low-level to matter?

Reply:
> If the node startup process detects existing TLS settings in the Elasticsearch keystore, the process halts and the node fails to start.

That's a good way to phrase it, I think.

The underlying reason to skip the auto-configuration in this case is similar to skipping it when the regular .yml TLS ones are encountered, ie. "it appears that the user already has configured Security to some extent".
An alternative suggestion would be to squeeze the keystore mention  in the point above 
> `xpack.security.transport.ssl.*` or`xpack.security.http.ssl.*`
settings have a value set (in the `elasticsearch.yml` or the `elasticsearch.keystore` files).